docs/decisions/0003-django-setting-toggles.rst

```diff
@@ -0,0 +1,66 @@
+------
+
+Draft
+
+Context
+-------
+
+Django Settings are implicitly defined when they are used.  For example, in ``edx-platform`` the ``FEATURES`` setting dict is used, and an example usage might look like::
+
+    settings.FEATURES.get('EXAMPLE_FEATURE', False)
+
+There are several problems with this style of implicit definition:
+
+* Each usage may or may not define a default value, and the default values are not guaranteed to be consistent.
+* It is unclear when and if a default value should be provided with the usage.
+* If there are multiple uses of a setting, it unclear where the code annotation that documents the toggle should live.
+* Django Settings toggles are unable to report the same level of detail in the toggle state report as Waffle toggles, because they can't be queried for location or code ownership information.
+
+Decision
+--------
+
+We will introduce a simple wrapping class that can be used to explicitly define a Django Setting toggle.
+
+Here is an example declaration::
+
+    EXAMPLE_FEATURE_TOGGLE = DjangoSettingToggle(
+        "FEATURES['EXAMPLE_FEATURE']",
+        default_value=False,
+    )
+
+It's usage can now match that of the waffle wrapping classes (e.g. WaffleFlag).
+
+Here is an example usage::
+
+    EXAMPLE_FEATURE_TOGGLE.is_enabled()
+
+DjangoSettingToggle will allow for expicit toggle definitions, with the following benefits:
+
+* A clear definition to annotate with documentation.
+* A clear place to set a consistent default value.
+* The ability to report in the toggle state endpoint the location and code ownership of any DjangoSettingToggle.
+* Additional linting would be possible to ensure manually created annotations are consistent with the code (e.g. default).
+
+Note: The method of using ``defaults.py`` for Django Setting defaults, as detailed in `OEP-45: Configuring and Operating Open edX`_, should only be used if the given IDA requires a different default from the one provided in code. The default provided in code should be the mostly widely used, or the safest from a security perspective, but may not be appropriate for all IDAs.
+
+.. _`OEP-45: Configuring and Operating Open edX`: https://open-edx-proposals.readthedocs.io/en/latest/oep-0045-arch-ops-and-config.html#configuration
+
+Consequences
+------------
+
+The main negative consequence is that developers would need to get used to a non-standard interface to working with Django Settings. Hopefully, the consistency with other toggle classes, like WaffleFlag and WaffleSwitch, should make this issue relatively trivial. Additionally, the decision to implement DjangoSettingToggle assumes the many advantages listed above outweigh this negative.
+
+Implementation Steps:
+
+* DjangoSettingToggle class needs to be implemented.
+* The new class should be added to the toggle state endpoint, including querying all instances (like WaffleFlag).
+* Any existing settings can be refactored to use the new class.
+
+Rejected Alternatives
+---------------------
```

Comment:
I'd like to propose a simpler alternative: that we remove all calls to `getattr(settings, "...", ...)` and `settings.FEATURES.get(...)`. Instead, make sure that all settings and features are properly defined in lms/envs/common.py and cms/envs/common.py. That way, we could simply write:

```
if settings.FEATURES["ENABLE_SOMEFEATURE"]:
    ...
```

or:

```
if settings.SOMESETTING:
    ...
```

I'm sure that with this solution we could also gain all the benefits listed above.

Reply:
This doesn’t keep the toggle definitions near the code. We want apps and modules to be more self-contained, and we have different teams that might be responsible for different apps, and I don’t see how your proposal addresses these issues.

Reply:
I understand your point, now that I've actually started documenting the feature toggles. The FEATURES setting is a long dict with many unsorted entries, which makes it difficult to realize which feature affects which part of the code.

The thing is: your proposed implementation makes it irrelevant to define default value for FEATURE entries, and even setting values (for feature toggles). Technically speaking this is not a problem, but it makes it difficult for operators to find settings that they can act on. Instead of grepping the FEATURES dict, they now have to grep the entire code base. This *might* not be an issue anymore once we finalize our work on feature toggle documentation, though.

If we do proceed with this approach, I think it's important that feature toggle defaults not be duplicated: both in the lms/envs/common.py and the toggles.py modules. Thus, we should strive to have an empty `FEATURES= {}` in common.py. This will pose a problem for feature toggles that have different default values in the LMS and the CMS.

Regarding the implementation: in order to avoid a call to `exec(...)` I would propose to write instead:

for toggles listed in FEATURES:

```
EXAMPLE_FEATURE_TOGGLE = DjangoSettingFeatureToggle(
    'EXAMPLE_FEATURE',
    default="somevalue"
)
```

for toggles defined by settings:

```
EXAMPLE_FEATURE_TOGGLE = DjangoSettingToggle(
    'EXAMPLE_SETTING',
    default="somevalue"
 )
```

Reply:
@robrap: Thanks for the explanations, this clarifies several things. Indeed, I wasn't aware of the new toggle state endpoint.

The only thing that I disagree on is the implementation:

> I was not planning on using exec, but parsing the string to walk the tree. I should update the example from FEATURES to EXAMPLE_SETTING_DICT. FEATURES is not the only dict that contains toggles.

This sounds a little brittle to me. I understand that `FEATURES` might not be the only dict to pull values from. Thus, I'd like to suggest the following implementation:

```
EXAMPLE_DICT_TOGGLE = DjangoSettingDictToggle(
    'EXAMPLE_SETTING_DICT',
    'EXAMPLE_DICT_KEY',
    default="somevalue"
)
```

Or equivalently:

```
EXAMPLE_DICT_TOGGLE = DjangoDictToggle(
    settings.EXAMPLE_SETTING_DICT,
    'EXAMPLE_DICT_KEY',
    default="somevalue"
)
```

Reply:
I'm not clear on what you mean by _brittle_.  Do you think it might unexpectedly break?  Do you think it just might not do what someone intends it to do?
Note: we have the option of making it work for `EXAMPLE_SETTING_DICT['EXAMPLE_DICT_KEY']` only if we want consistency, but equally could make it work for `EXAMPLE_SETTING_DICT["EXAMPLE_DICT_KEY"]` and even `EXAMPLE_SETTING_DICT[EXAMPLE_DICT_KEY]` if we want to err on making it flexibly accept all versions someone might try. This also has the advantage of a single class for all the settings.

I can get a couple of other opinions as well.  I don't think this particular decision makes or breaks anything regarding the overall intent of the ADR.

Reply:
> I'm not clear on what you mean by brittle? Do you think it might unexpectedly break? Do you think it just might not do what someone intends it to do?

My point is that we would have to write a regular expression that would parse the string argument, and we will have to take into account quite a few edge cases, like the ones you described. On the other hand, we can write a one-liner for the following implementation:

```
def DjangoSettingDictToggle(setting_dict_name, key, default=False):
    return getattr(settings, setting_dict_name, {}).get(key, default=default)
```

(by the way: should this be `django_setting_dict_toggle`, as it is a function, and not a class?)

> I can get a couple of other opinions as well. I don't think this particular decision makes or breaks anything regarding the overall intent of the ADR.

Absolutely. It's an implementation detail.

Reply:
@regisb 
1. I moved my preferred version of your `DjangoSettingDictToggle` example to the ADR, and added a TODO to follow-up.
2. This should be a class and not a method because the class tracks its instances.  It also makes it more consistent to declare and use as other toggles.
3. If and when you are ready to thumb this, I'd like to land this ADR in draft mode, so people can review without feeling like they need to get up to date with all this history.  When we discuss implementation for dicts, we can refer to any relevant discussion.
Please give me a thumb if you think this is close enough to land as a Draft.

Reply:
:ok_hand: